provider/kubernetes/client.go

```diff
@@ -198,32 +195,30 @@ func (c *clientImpl) UpdateIngressStatus(namespace, name, ip, hostname string) e
 
 // GetService returns the named service from the given namespace.
 func (c *clientImpl) GetService(namespace, name string) (*corev1.Service, bool, error) {
-	var service *corev1.Service
-	item, exists, err := c.factories[c.lookupNamespace(namespace)].Core().V1().Services().Informer().GetStore().GetByKey(namespace + "/" + name)
-	if item != nil {
-		service = item.(*corev1.Service)
+	service, err := c.factories[c.lookupNamespace(namespace)].Core().V1().Services().Lister().Services(namespace).Get(name)
+	if err != nil {
+		return nil, !kubeerror.IsNotFound(err), err
 	}
-	return service, exists, err
+	return service, true, nil
 }
 
 // GetEndpoints returns the named endpoints from the given namespace.
 func (c *clientImpl) GetEndpoints(namespace, name string) (*corev1.Endpoints, bool, error) {
-	var endpoint *corev1.Endpoints
-	item, exists, err := c.factories[c.lookupNamespace(namespace)].Core().V1().Endpoints().Informer().GetStore().GetByKey(namespace + "/" + name)
-	if item != nil {
-		endpoint = item.(*corev1.Endpoints)
+	endpoint, err := c.factories[c.lookupNamespace(namespace)].Core().V1().Endpoints().Lister().Endpoints(namespace).Get(name)
+	if err != nil {
+		return nil, !kubeerror.IsNotFound(err), err
```

Comment:
IIUC, we will now return a non-nil error even if this is just about a missing element. However, our [call site](https://github.com/yue9944882/traefik/blob/402ef53199819f6eee1d03fc00430f684dde9e2e/provider/kubernetes/kubernetes.go#L302) still considers non-nil errors as triggers to end Ingress parsing prematurely and back off as there is some presumed issue with the Kubernetes API connection.

I think we still need to translate _not found_ errors into a `nil, false, nil` triple (or move the distinction to the call site).

Similar for other changes you have made.

Reply:
> (or move the distinction to the call site)

In order to narrow the scope of changes, I'd prefer the former solution. 😀